### PR TITLE
Feature/logout link on rpc error

### PR DESCRIPTION
**Pull request by GCRev**
See: https://github.com/rstudio/rstudio-pro/issues/5646

### Intent

image

Partially address the original issue by displaying a different warning message and a link to attempt to log in again.

### Approach

After a bit of discussion with Gary, we decided that it would be acceptable to take a bit of a shortcut, because the alternative would require us to touch upwards of 120 files. As this is an edge case, the rest of the UI is already inoperable.

### Automated Tests

At the moment, testing this using automation is infeasible because we do not have a way to force-simulate a status code of 0 in response to an RPC request. Testing at all is tricky because either the original condition must be met, which is related to SAML expiration, or the server has to be modified in order to return a 0 status.

### QA Notes

Presently, I'm not sure what a reasonable approach is to test this, even manually.

### Documentation

@AshleyHenry15 do you think this bears mentioning in our docs somewhere?

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

File: src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java [hunk 1]
```diff
@@ -169,18 +171,47 @@ public void onResponseReceived(Request request,
...
+                     {
+                        // only show this error if not desktop and return early. Otherwise handle error as normal
+                        message = constants_.rpcOverrideErrorMessageServer((Desktop.isDesktop() ? constants_.rSessionMessage() : constants_.rStudioServerMessage()));
+                     
+                        if (!showingNoConnectError_) 
+                        {
+                           showingNoConnectError_ = true;
+                           RStudioGinjector.INSTANCE.getGlobalDisplay().showMessage(
+                              GlobalDisplay.MSG_ERROR,
+                              "RPC Error",
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java:195]: Should add this to the `CoreClientConstants` string table so it can be localized.

File: src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java [hunk 2]
```diff
@@ -65,7 +65,7 @@ public void onError(ServerError error)
                message = errMsg.stringValue();
          }
       }
-      
+
```
Comment (gtritchie) [inline src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java:68]: nit: unnecessary whitespace change

**Review by gtritchie — CHANGES_REQUESTED** [review]
A couple minor things but overall looking good.

File: src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java (hunk 2 above)
Comment (gtritchie) [reply src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java:68]: Unless I also made that change in main, in which case it's fine :-)

File: src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java [hunk 3]
```diff
@@ -169,18 +171,47 @@ public void onResponseReceived(Request request,
...
                   {
-                     message = constants_.rpcOverrideErrorMessage((Desktop.isDesktop() ? constants_.rSessionMessage() : constants_.rStudioServerMessage()), getMethod());
+                     type = RpcError.TRANSMISSION_ERROR_NO_RESPONSE;
+                     if (Desktop.isDesktop())
+                     {
+                        message = constants_.rpcOverrideErrorMessage((Desktop.isDesktop() ? constants_.rSessionMessage() : constants_.rStudioServerMessage()), getMethod());
+                     } 
+                     else
+                     {
+                        // only show this error if not desktop and return early. Otherwise handle error as normal
```
[the remark this answers is not in the document]
Comment (GCRev) [reply src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java:187]: I feel like it's fine to show this in any server context, since there is a concept of logging in or out even in the OS server (unless there isn't).

File: src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java (hunk 1 above)
Comment (GCRev) [reply src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java:195]: Will do.

File: src/gwt/src/org/rstudio/core/client/MessageDisplay.java [hunk 4]
```diff
@@ -196,6 +196,37 @@ public void showMessage(int type,
...
+
+   public void showMessage(int type,
+                           String caption,
+                           String message,
+                           String urlLabel,
+                           String url,
+                           Operation dismissed)
+   {
+      createDialog(type, caption, message)
+         .addLink(urlLabel, url)
```
Comment (jeffvroom) [inline src/gwt/src/org/rstudio/core/client/MessageDisplay.java:225]: Is this link going to have target="_blank"?  We need that link to open up in a new tab or else the user will lose work in their session.  Also, I think the message should say: "Please try to login in a new tab and return here to continue the session".

File: src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java (hunk 2 above)
Comment (GCRev) [reply src/gwt/src/org/rstudio/studio/client/common/SimpleRequestCallback.java:68]: I probably put that in at some point when I was experimenting, and I generally prefer whitespace after closing brackets. I can take this back out.

File: src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gtritchie) [reply src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java:187]: Open source server has a login page, and a logout widget in the IDE, same as pro. Just installed ocean-storm server to make sure I'm not crazy:

Screenshot 2024-01-17 at 16 15 54

Screenshot 2024-01-17 at 16 17 39

File: src/gwt/src/org/rstudio/core/client/MessageDisplay.java (hunk 4 above)
Comment (gtritchie) [reply src/gwt/src/org/rstudio/core/client/MessageDisplay.java:225]: Yes, per the comment: "The link will open in a new tab (e.g. target=_blank)." That happens in MessageDialog.java (also in this review) via:

```java
Anchor anchor = new Anchor(label, url, "_blank");
```

File: src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java [hunk 5]
```diff
@@ -169,18 +171,47 @@ public void onResponseReceived(Request request,
...
+                        message = constants_.rpcOverrideErrorMessageServer((Desktop.isDesktop() ? constants_.rSessionMessage() : constants_.rStudioServerMessage()));
```
[the remark this answers is not in the document]
Comment (GCRev) [reply src/gwt/src/org/rstudio/core/client/jsonrpc/RpcRequest.java:188]: Will change.

File: src/gwt/src/org/rstudio/core/client/MessageDisplay.java (hunk 4 above)
Comment (GCRev) [reply src/gwt/src/org/rstudio/core/client/MessageDisplay.java:225]: I'll modify the message to make it clearer that they should attempt to re-auth without closing everything out, since it should fix the RPC issues to do that.

Comment (GCRev) [reply src/gwt/src/org/rstudio/core/client/MessageDisplay.java:225]: Re-worded to

> Unable to establish connection with session on {0}. Please click the link to try logging in again in a new tab, then return to resume your session.

**Comment by jeffvroom** [comment]
LGTM!

**Review by jeffvroom — APPROVED** [review]

**Review by gtritchie — APPROVED** [review]
LGTM!
